lib/MetadataStore.js

```diff
@@ -0,0 +1,380 @@
+          yield this._conn.execute(migration);
+        }
+      }.bind(this));
+    } catch (e) {
+      Cu.reportError("MetadataStore failed to create tables.");
+      throw e;
+    }
+  }),
+
+  get transactionInProgress() {
+    return this._conn.transactionInProgress;
+  },
+
+  /**
+   * Creates a connection to the metadata database. It sets the journal mode
+   * to WAL, enables the foreign key support, and also creates the tables and
+   * indices if necessary
+   *
+   * Returns a promise that is resolved upon success, or rejected if an exception occurs
+   */
+  asyncConnect: Task.async(function*() {
+    if (this._conn) {
+      return;
+    }
+
+    try {
+      this._conn = yield Sqlite.openConnection({path: this._path});
+      yield this._conn.execute("PRAGMA journal_mode = WAL;");
+      yield this._conn.execute("PRAGMA foreign_keys = ON;");
+      yield this._asyncCreateTableSchema();
+    } catch (e) {
+      Cu.reportError("MetadataStore failed to create connection: " + e.message);
+      throw e;
+    }
+  }),
+
+  asyncClose: Task.async(function*() {
+    if (this._conn) {
+      yield this._conn.close();
+      this._conn = null;
+    }
+  }),
+
+  _getMetadataParameters(aRow) {
+    return {
+      cache_key: aRow.cache_key,
+      canonical_url: aRow.canonical_url,
+      title: aRow.title,
+      type: aRow.type || null,
+      description: aRow.description,
+      media_url: aRow.media.url || null,
+      expired_at: aRow.expired_at
+    };
+  },
+
+  _getFaviconParameters(aRow) {
+    return {
+      url: aRow.favicon_url,
+      type: IMAGE_TYPES.favicon,
+      height: 0,
+      width: 0,
+      color: this._rgbToHex(aRow.favicon_colors[0].color)
+    };
+  },
+
+  _getImageParameters(aRow) {
+    return {
+      url: aRow.url,
+      type: IMAGE_TYPES.preview,
+      height: aRow.height,
+      width: aRow.width,
+      color: this._rgbToHex(aRow.colors[0].color)
+    };
+  },
+
+  /**
+  * Convert a RGB array to the Hex form, e.g. [10, 1, 2] => #0A0102
+  */
+  _rgbToHex(rgb) {
+    return "#" + ((1 << 24) + (rgb[0] << 16) + (rgb[1] << 8) + rgb[2]).toString(16).slice(1).toUpperCase();
+  },
+
+  _asyncGetLastInsertRowID: Task.async(function*() {
+    let result = yield this._conn.execute(SQL_LAST_INSERT_ROWID);
+    return result[0].getResultByName("lastInsertRowID");
+  }),
+
+  _asyncGetImageIDByURL: Task.async(function*(url) {
+    let result = yield this._conn.executeCached(SQL_SELECT_IMAGE_ID, [url]);
+    return result[0].getResultByName("id");
+  }),
+
+  /**
+   * Inserts the metadata into the database. It consists of two tables,
+   * i.e. page_metadata for the regular meta information of the page, and
+   * page_images for the favicon and preview images.
+   *
+   * aRows, an array of metadata objects that currently generated by embed.ly
+   * Returns a promise that is resolved upon success, or rejected if an exception occurs
+   */
+  asyncInsert: Task.async(function*(aRows) {
```

Comment:
it's confusing to call this parameter `aRows`. perhaps `metaObjs` makes more sense. `aRows` gives the impression these are database rows, which they aren't.